File: fsharp/AoC2021/Day02/Day02/Solution.fs
module Solution

open Instruction

let folder (depth: int, position: int) (instruction: Instruction) =
    match instruction with
    | Forward x -> (depth, position + x)
    | Down x -> (depth + x, position)
    | Up x -> (depth - x, position)

let computeDepthAndPosition (instructions: seq<Instruction>) = Seq.fold folder (0, 0) instructions

let folderWithAim (depth: int, position: int, aim: int) (instruction: Instruction) =
    match instruction with
    | Forward x -> (depth + x * aim, position + x, aim)
    | Down x -> (depth, position, aim + x)
    | Up x -> (depth, position, aim - x)

let computeDepthAndPositionWithAim (instructions: seq<Instruction>) =
    let depth, position, _ =
        Seq.fold folderWithAim (0, 0, 0) instructions

    (depth, position)

File: csharp/AoC2021/Day02/Day02/Solution.cs
using System.Diagnostics.CodeAnalysis;

namespace Day02;

public static class Solution
{
    [ExcludeFromCodeCoverage]
    public static void Solve(IEnumerable<string> input)
    {
        var instructions = input
            .Select(Instruction.Parse)
            .ToList();

        var (depth1, position1) = FollowInstructions(instructions);
        Console.WriteLine($"Not considering aim, the submarine ends at position {position1} and depth {depth1}, " +
                          $"for a final answer of {position1 * depth1}.");

        var (depth2, position2) = FollowInstructionsWithAim(instructions);
        Console.WriteLine($"Considering aim, the submarine ends at position {position2} and depth {depth2}, " +
                          $"for a final answer of {position2 * depth2}.");
    }

    public static (int, int) FollowInstructions(List<Instruction> instructions)
    {
        var depth = 0;
        var position = 0;

        foreach (var instruction in instructions)
        {
            switch (instruction.Direction)
            {
                case Direction.Forward:
                    position += instruction.Distance;
                    break;
                case Direction.Up:
                    depth -= instruction.Distance;
                    break;
                case Direction.Down:
                    depth += instruction.Distance;
                    break;
            }
        }

        return (depth, position);
    }

    public static (int, int) FollowInstructionsWithAim(List<Instruction> instructions)
    {
        var aim = 0;
        var depth = 0;
        var position = 0;

        foreach (var instruction in instructions)
        {
            switch (instruction.Direction)
            {
                case Direction.Forward:
                    position += instruction.Distance;
                    depth += instruction.Distance * aim;
                    break;
                case Direction.Up:
                    aim -= instruction.Distance;
                    break;
                case Direction.Down:
                    aim += instruction.Distance;
                    break;
            }
        }

        return (depth, position);
    }
}
